File: templates/wasm-node-cpp/examples/http_request.cpp
/// HTTP Request Node — Demonstrates declaring HTTP permissions (C++)
///
/// This example shows how to declare the "http" permission and use the
/// WIT-generated host import to make outbound HTTP requests from a C++ WASM node.
/// Copy this pattern into your node.cpp when you need network access.
///
/// With the Component Model you also get native WASI sockets support for
/// direct TCP/UDP/DNS — but for simple HTTP, the host-provided request
/// function is the easiest route.

#include <cstdint>
#include <cstring>
#include <string>
#include <vector>

#include "flow_like_node.h"

// ── Helpers (same as in node.cpp — in a real project, share via a header) ───

static flow_like_node_string_t to_wit_string(const std::string& s) {
    flow_like_node_string_t ws;
    ws.ptr = reinterpret_cast<uint8_t*>(const_cast<char*>(s.data()));
    ws.len = s.size();
    return ws;
}

static std::string from_wit_string(const flow_like_node_string_t* ws) {
    if (!ws || !ws->ptr || ws->len == 0) return "";
    return std::string(reinterpret_cast<const char*>(ws->ptr), ws->len);
}

static std::string json_quote(const std::string& s) {
    std::string out;
    out.reserve(s.size() + 2);
    out += '"';
    for (char c : s) {
        switch (c) {
            case '"':  out += "\\\""; break;
            case '\\': out += "\\\\"; break;
            case '\n': out += "\\n";  break;
            case '\r': out += "\\r";  break;
            case '\t': out += "\\t";  break;
            default:   out += c;      break;
        }
    }
    out += '"';
    return out;
}

// ── HTTP GET via WIT import ─────────────────────────────────────────────
//
// The WIT http interface provides:
//   request(method: u8, url: string, headers: string, body: option<list<u8>>) -> option<string>
//
// method: 0 = GET, 1 = POST, 2 = PUT, 3 = DELETE, 4 = PATCH
//
// Example usage:

static std::string http_get(const std::string& url, const std::string& headers_json) {
    flow_like_node_string_t w_url     = to_wit_string(url);
    flow_like_node_string_t w_headers = to_wit_string(headers_json);

    // No body for GET
    flow_like_node_option_list_u8_t no_body;
    no_body.is_some = false;

    flow_like_node_option_string_t result;
    flow_like_node_http_request(0, &w_url, &w_headers, &no_body, &result);

    if (result.is_some) {
        return from_wit_string(&result.val);
    }
    return "";
}

/// Build the node definition for this example.
/// Note: add_permission("http") is required for the host to allow requests.
///
/// In your node.cpp, the build_http_get_definition() return value would be
/// serialised by exports_flow_like_node_get_node / get_nodes.
///
/// This file is not compiled as-is; it is reference code you copy into your
/// node.cpp.

// static NodeDefinition build_http_get_definition() {
//     NodeDefinition def;
//     def.name          = "http_get_request_cpp";
//     def.friendly_name = "HTTP GET Request (C++)";
//     def.description   = "Sends a GET request to a URL and reports the result";
//     def.category      = "Network/HTTP";
//     def.add_permission("http");
//
//     def.add_pin(PinDefinition::input("exec", "Execute", "Trigger", "Exec"));
//     def.add_pin(PinDefinition::input("url", "URL", "Target URL", "String")
//                     .with_default("\"https://httpbin.org/get\""));
//     def.add_pin(PinDefinition::input("headers_json", "Headers", "JSON headers", "String")
//                     .with_default("\"{}\""));
//     def.add_pin(PinDefinition::output("exec_out", "Done", "Fires after request", "Exec"));
//     def.add_pin(PinDefinition::output("response", "Response", "Response body", "String"));
//     return def;
// }

File: templates/wasm-node-zig/examples/http_request.zig
/// HTTP Request Node — Demonstrates declaring HTTP permissions (Zig, Component Model)
///
/// This example shows how to declare the "http" permission and use the
/// WIT-generated host imports to make outbound HTTP requests from a Zig WASM node.
/// Copy this pattern into your main.zig when you need network access.

const std = @import("std");
const wit = @cImport({
    @cInclude("flow_like_node.h");
});

const allocator = std.heap.page_allocator;

// ============================================================================
// Helpers (same as main.zig)
// ============================================================================

fn toWitString(s: []const u8) wit.flow_like_node_string_t {
    return .{
        .ptr = @constCast(@ptrCast(s.ptr)),
        .len = s.len,
    };
}

fn fromWitString(ws: *const wit.flow_like_node_string_t) []const u8 {
    if (ws.ptr == null or ws.len == 0) return "";
    const p: [*]const u8 = @ptrCast(ws.ptr.?);
    return p[0..ws.len];
}

fn setWitResult(src: []const u8, ret: *wit.flow_like_node_string_t) void {
    const buf: [*]u8 = @ptrCast(wit.cabi_realloc(
        null,
        0,
        1,
        src.len,
    ) orelse return);
    @memcpy(buf[0..src.len], src);
    ret.ptr = buf;
    ret.len = src.len;
}

fn buildPinJson(name: []const u8, friendly_name: []const u8, description: []const u8, pin_type: []const u8, data_type: []const u8, default_value: ?[]const u8) []const u8 {
    const default_part = if (default_value) |dv|
        std.fmt.allocPrint(allocator, ",\"default_value\":{s}", .{dv}) catch ""
    else
        "";
    return std.fmt.allocPrint(allocator,
        "{{\"name\":\"{s}\",\"friendly_name\":\"{s}\",\"description\":\"{s}\",\"pin_type\":\"{s}\",\"data_type\":\"{s}\"{s}}}",
        .{ name, friendly_name, description, pin_type, data_type, default_part },
    ) catch "{}";
}

// ============================================================================
// Node definition — note "http" permission
// ============================================================================

fn buildHttpGetDefinition() []const u8 {
    const pins = [_][]const u8{
        buildPinJson("exec", "Execute", "Trigger execution", "Input", "Exec", null),
        buildPinJson("url", "URL", "Target URL", "Input", "String", "\"\\\"https://httpbin.org/get\\\"\""),
        buildPinJson("headers_json", "Headers (JSON)", "Request headers as JSON", "Input", "String", "\"\\\"{}\\\"\""),
        buildPinJson("exec_out", "Done", "Fires after the request", "Output", "Exec", null),
        buildPinJson("success", "Success", "Whether the HTTP call was accepted", "Output", "Bool", null),
    };

    var pins_json = std.ArrayList(u8).init(allocator);
    pins_json.appendSlice("[") catch {};
    for (pins, 0..) |p, i| {
        if (i > 0) pins_json.appendSlice(",") catch {};
        pins_json.appendSlice(p) catch {};
    }
    pins_json.appendSlice("]") catch {};

    return std.fmt.allocPrint(allocator,
        "{{\"name\":\"http_get_request_zig\",\"friendly_name\":\"HTTP GET Request (Zig)\",\"description\":\"Sends a GET request to a URL and reports the result\",\"category\":\"Network/HTTP\",\"pins\":{s},\"long_running\":false,\"abi_version\":1,\"permissions\":[\"http\"]}}",
        .{pins_json.items},
    ) catch "{}";
}

// ============================================================================
// Run handler
// ============================================================================

fn handleHttpGet() []const u8 {
    // Read inputs via WIT pins interface
    var url_name = toWitString("url");
    var url_opt: wit.flow_like_node_option_string_t = undefined;
    wit.flow_like_node_pins_get_input(&url_name, &url_opt);
    const url = if (url_opt.is_some) fromWitString(&url_opt.val) else "https://httpbin.org/get";

    var headers_name = toWitString("headers_json");
    var headers_opt: wit.flow_like_node_option_string_t = undefined;
    wit.flow_like_node_pins_get_input(&headers_name, &headers_opt);
    const headers = if (headers_opt.is_some) fromWitString(&headers_opt.val) else "{}";

    var log_msg = toWitString("Sending GET request");
    wit.flow_like_node_logging_log(1, &log_msg);

    // Method 0 = GET
    var wurl = toWitString(url);
    var wheaders = toWitString(headers);
    const no_body = wit.flow_like_node_option_list_u8_t{ .is_some = false, .val = undefined };
    var result_opt: wit.flow_like_node_option_string_t = undefined;
    wit.flow_like_node_http_request(0, &wurl, &wheaders, &no_body, &result_opt);

    const ok = result_opt.is_some;
    const ok_str = if (ok) "true" else "false";

    var out_name = toWitString("success");
    var out_val = toWitString(ok_str);
    wit.flow_like_node_pins_set_output(&out_name, &out_val);

    var exec_pin = toWitString("exec_out");
    wit.flow_like_node_pins_activate_exec(&exec_pin);

    return std.fmt.allocPrint(allocator,
        "{{\"outputs\":{{\"success\":{s}}},\"activate_exec\":[\"exec_out\"],\"pending\":false}}",
        .{ok_str},
    ) catch "{}";
}

// ============================================================================
// WIT exports — wire up to the http_get node
// ============================================================================

// NOTE: This is an example — in a real multi-node package you would dispatch
// based on some identifier. For a single-node package, replace the exports
// in main.zig with the patterns shown here.
